libwnbd/wnbd_ioctl.cpp

```diff
@@ -139,6 +144,367 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
     return Status;
 }
 
+DWORD RemoveWnbdInf(_In_ LPCTSTR InfName)
+{
+    DWORD Status = 0;
+    HINF HandleInf = INVALID_HANDLE_VALUE;
+    TCHAR InfData[MAX_INF_STRING_LENGTH];
+    UINT ErrorLine;
+    INFCONTEXT Context;
+    BOOLEAN Found = FALSE;
+
+    HandleInf = SetupOpenInfFile(InfName, NULL, INF_STYLE_WIN4, &ErrorLine);
+    if (HandleInf == INVALID_HANDLE_VALUE) {
+        Status = GetLastError();
+        LogError("SetupOpenInfFile failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+    if (!SetupFindFirstLine(HandleInf, INFSTR_SECT_VERSION, INFSTR_KEY_CATALOGFILE, &Context)) {
+        Status = GetLastError();
+        LogError("SetupFindFirstLine failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+    if (!SetupGetStringField(&Context, 1, InfData, ARRAYSIZE(InfData), NULL)) {
+        Status = GetLastError();
+        LogError("SetupGetStringField failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+
+    /* Match the OEM inf file based on the catalog string wnbd.cat */
+    if (!wcscmp(InfData, L"wnbd.cat")) {
+        std::wstring SearchString(InfName);
+        if (!SetupUninstallOEMInf(
+            SearchString.substr(SearchString.find_last_of(L"\\") + 1).c_str(), SUOI_FORCEDELETE, 0)) {
+            Status = GetLastError();
+            LogError("SetupUninstallOEMInfA failed with "
+                "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        }
+    }
+
+failed:
+    if (HandleInf != INVALID_HANDLE_VALUE) {
+        SetupCloseInfFile(HandleInf);
+    }
+
+    return Status;
+}
+
+/* Cycle through all OEM inf files and try a best effort mode to remove all
+ * files which include the string wnbd.cat */
+VOID CleanDrivers()
+{
+    DWORD Status = 0;
+    TCHAR OemName[MAX_PATH];
+    HANDLE DirHandle = INVALID_HANDLE_VALUE;
+    WIN32_FIND_DATA OemFindData;
+
+    if (!GetWindowsDirectory(OemName, ARRAYSIZE(OemName))) {
+        Status = GetLastError();
+        LogError("GetWindowsDirectory failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return;
+    }
+    if (wcscat_s(OemName, ARRAYSIZE(OemName), L"\\INF\\OEM*.INF")) {
```

Comment:
Instead of parsing all *.inf files, isn't it better to just use the ``SetupDi`` API to retrieve the right one, like you did before? https://github.com/cloudbase/wnbd/commit/57f523285da54f8963c4f5f7a5048466a1803866#diff-0a246dbc425e428f1d11ae42c36d302a5fe6ed52d3c3a3e499a49e565752256dR140

Reply:
That would be wonderful :), but unfortunately `SetupDi` works only on registered devices and not the driver catalog. Same applies for the `Di` APIs.

From what I found out it is either parsing all the OEM*.inf files or cycling through all the driver registry entries.